### PR TITLE
Basic support for memory initialization checks for unions

### PR DESCRIPTION
**Pull request by artemagvanian**
This PR introduces very basic support for memory initialization checks for unions.

As of now, the following is supported:
- Unions can be created
- Union fields can be assigned to
- Union fields can be read from
- Unions can be assigned directly to another union

For more information about planned functionality, see #3300 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: tests/expected/uninit/unions.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+/// Tests accessing uninit data via subfields of a union.
+#[kani::proof]
+unsafe fn union_complex_subfields() {
+    let u = U1 { a: (0, 0) };
+    let non_padding = u.b.1;
+}
+
+/// Tests overwriting data inside unions.
+#[kani::proof]
+unsafe fn union_update() {
```
Comment (celinval) [inline tests/expected/uninit/unions.rs:47]: can you please add `should_fail` and `should_pass` suffix to this test and the one above?

File: tests/expected/uninit/fixme_unions.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+#[repr(C)]
+#[derive(Clone, Copy)]
+union U {
+    a: u16,
+    b: u32,
+}
+
+/// Reading padding data via simple union access if union is passed to another function.
+#[kani::proof]
+unsafe fn cross_function_union() {
```
Comment (celinval) [inline tests/expected/uninit/fixme_unions.rs:18]: Can you please add `should_pass` / `should_fail` as suffix, or update the comment to state what the expectation is? Just to make it easier to debug later on. Thanks!

**Review by celinval — COMMENTED** [review]
Just reviewed the tests and the library changes for now since the rest will likely be simpler after your other PR is merged. Thanks!

File: kani-compiler/src/kani_middle/transform/check_uninit/mod.rs [hunk 3]
```diff
@@ -426,14 +496,15 @@ impl<'a, 'tcx> UninitInstrumenter<'a, 'tcx> {
...
         let copy_init_state_instance = resolve_mem_init_fn(
             get_mem_init_fn_def(self.tcx, "KaniCopyInitState", &mut self.mem_init_fn_cache),
-            layout.len(),
+            layout_size,
             *pointee_info.ty(),
         );
         let position = operation.position();
-        let MemoryInitOp::Copy { from, to, count } = operation else { unreachable!() };
+        let (from, to) = operation.expect_copy_operands();
+        let count = operation.expect_count();
```
Comment (artemagvanian) [inline kani-compiler/src/kani_middle/transform/check_uninit/mod.rs:507]: Those changes are due to extracting some common functionality between `build_copy` and `build_assign_union`

File: tests/expected/uninit/unions.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+unsafe fn union_complex_subfields_should_fail() {
+    let u = U1 { a: (0, 0) };
+    let non_padding = u.b.1;
+}
+
+/// Tests accessing uninitialized data inside a place projection.
+#[kani::proof]
+unsafe fn union_uninit_inside_projection_should_fail() {
+    let u = U1 { a: (0, 0) };
+    let non_padding = u.b.0;
```
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply tests/expected/uninit/unions.rs:49]: Is it because `u.b` itself accesses uninitialized data?

File: tests/expected/uninit/unions.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+/// Tests overwriting data inside unions.
+#[kani::proof]
+unsafe fn union_update_should_pass() {
+    let mut u = U { a: 0 };
+    u.b = 0;
+    let non_padding = u.b;
```
Comment (zhassan-aws) [inline tests/expected/uninit/unions.rs:57]: Add an assert.

File: tests/expected/uninit/fixme_unions.rs [hunk 6]
```diff
@@ -0,0 +1,60 @@
+// Copyright Kani Contributors
```
Comment (zhassan-aws) [inline tests/expected/uninit/fixme_unions.rs:1]: Is this test intentionally missing an expected file? If so, please add a TODO comment to add an expected file and rename the test once the necessary support is added.

**Review by zhassan-aws — COMMENTED** [review]
Nice!

File: tests/expected/uninit/fixme_unions.rs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        let padding = u.b; // Read 32 bytes from `u`.
+    }
+    let u = U { a: 0 }; // `u` is initialized for 16 bytes.
+    helper(u);
+}
+
+/// Reading padding data but a union is via a union from behind a pointer.
+#[kani::proof]
+unsafe fn pointer_union_should_fail() {
+    let u = U { a: 0 }; // `u` is initialized for 16 bytes.
```
[the remark this answers is not in the document]
Comment (artemagvanian) [reply tests/expected/uninit/fixme_unions.rs:29]: 🙈

File: library/kani/src/mem_init.rs [hunk 8]
```diff
@@ -321,3 +321,16 @@ fn copy_init_state<const LAYOUT_SIZE: usize, T>(from: *const T, to: *const T, nu
         MEM_INIT_STATE.copy::<LAYOUT_SIZE>(from_ptr as *const u8, to_ptr as *const u8, num_elts);
     }
 }
+
+/// Copy initialization state of `size_of::<T>` bytes from one pointer to the other.
```
[the remark this answers is not in the document]
Comment (artemagvanian) [reply library/kani/src/mem_init.rs:325]: Yes, that's correct. `LAYOUT_SIZE` is not strictly necessary here, but I kept it in the signature to keep the generics the same as in the other instrumentation functions. I'll add a comment stating that `LAYOUT_SIZE == size_of::<T>`.

File: library/kani/src/mem_init.rs [hunk 9]
```diff
@@ -321,3 +321,16 @@ fn copy_init_state<const LAYOUT_SIZE: usize, T>(from: *const T, to: *const T, nu
...
+#[rustc_diagnostic_item = "KaniCopyInitStateSingle"]
+fn copy_init_state_single<const LAYOUT_SIZE: usize, T>(from: *const T, to: *const T) {
+    if LAYOUT_SIZE == 0 {
+        return;
+    }
+    let (from_ptr, _) = from.to_raw_parts();
```
[the remark this answers is not in the document]
Comment (artemagvanian) [reply library/kani/src/mem_init.rs:331]: Not necessary, just to keep it similar with other instrumentation functions.

File: library/kani/src/mem_init.rs [hunk 10]
```diff
@@ -321,3 +321,16 @@ fn copy_init_state<const LAYOUT_SIZE: usize, T>(from: *const T, to: *const T, nu
         MEM_INIT_STATE.copy::<LAYOUT_SIZE>(from_ptr as *const u8, to_ptr as *const u8, num_elts);
     }
 }
+
+/// Copy initialization state of `size_of::<T>` bytes from one pointer to the other.
+#[rustc_diagnostic_item = "KaniCopyInitStateSingle"]
+fn copy_init_state_single<const LAYOUT_SIZE: usize, T>(from: *const T, to: *const T) {
```
[the remark this answers is not in the document]
Comment (artemagvanian) [reply library/kani/src/mem_init.rs:327]: Correct -- thanks!

File: tests/expected/uninit/unions.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (artemagvanian) [reply tests/expected/uninit/unions.rs:49]: I just got a response from `t-opsem`, and you are right -- this is actually valid. I will make changes to account for that

File: tests/expected/uninit/fixme_unions.rs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+#[kani::proof]
+unsafe fn cross_function_union_should_fail() {
+    unsafe fn helper(u: U) {
+        let padding = u.b; // Read 4 bytes from `u`.
+    }
+    let u = U { a: 0 }; // `u` is initialized for 2 bytes.
+    helper(u);
+}
+
+/// Reading padding data but a union is via a union from behind a pointer.
```
Comment (zhassan-aws) [inline tests/expected/uninit/fixme_unions.rs:27]: Cannot parse this comment :)

File: kani-compiler/src/kani_middle/transform/check_uninit/mod.rs [hunk 12]
```diff
@@ -426,14 +497,15 @@ impl<'a, 'tcx> UninitInstrumenter<'a, 'tcx> {
             local: body.new_local(Ty::new_tuple(&[]), source.span(body.blocks()), Mutability::Not),
             projection: vec![],
         };
-        let PointeeLayout::Sized { layout } = pointee_info.layout() else { unreachable!() };
+        let layout_size = pointee_info.layout().maybe_size().unwrap();
         let copy_init_state_instance = resolve_mem_init_fn(
             get_mem_init_fn_def(self.tcx, "KaniCopyInitState", &mut self.mem_init_fn_cache),
```
Comment (zhassan-aws) [inline kani-compiler/src/kani_middle/transform/check_uninit/mod.rs:502]: const str for diagnostic function.

File: kani-compiler/src/kani_middle/transform/check_uninit/mod.rs [hunk 13]
```diff
@@ -443,6 +515,45 @@ impl<'a, 'tcx> UninitInstrumenter<'a, 'tcx> {
...
+        pointee_info: PointeeInfo,
+    ) {
+        let ret_place = Place {
+            local: body.new_local(Ty::new_tuple(&[]), source.span(body.blocks()), Mutability::Not),
+            projection: vec![],
+        };
+        let mut statements = vec![];
+        let layout_size = pointee_info.layout().maybe_size().unwrap();
+        let copy_init_state_instance = resolve_mem_init_fn(
+            get_mem_init_fn_def(self.tcx, "KaniCopyInitStateSingle", &mut self.mem_init_fn_cache),
```
Comment (zhassan-aws) [inline kani-compiler/src/kani_middle/transform/check_uninit/mod.rs:533]: Same here

**Review by zhassan-aws — APPROVED** [review]
LGTM. Only a few nits.

File: tests/expected/uninit/fixme_unions.rs (hunk 11 above)
Comment (artemagvanian) [reply tests/expected/uninit/fixme_unions.rs:27]: Totally fair :) "via a union from" should not have been there

**Pull request merged**